README.md
# hospital-schemas

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2021 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-11-15
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-11-15
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/hospital-schemas/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/hospital-schemas/compare/0.1.0...0.1.1


test/hospital_schemas/core_test.clj
(ns hospital-schemas.core-test
  (:require [clojure.test :refer :all]
            [hospital-schemas.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/hospital_schemas/aula1.clj
(ns hospital_schemas.aula1
  (:use clojure.pprint)
  (:require [schema.core :as s]))

(defn adiciona-paciente
  [pacientes, paciente]
  (if-let [id (:id paciente)]
    (assoc pacientes id paciente)
    (throw (ex-info "Paciente não possui id" {:paciente paciente}))))

; { 15 [], 20 [], 24 [] }
(defn adiciona-visitas
  [visitas, paciente, novas-visitas]
  (if (contains? visitas paciente)
    (update visitas paciente concat novas-visitas)
    (assoc visitas paciente novas-visitas)))

(defn imprime-relatorio-de-paciente [visitas, paciente]
  (println "Visitas do paciente" paciente "são" (get visitas paciente)))

(defn testa-uso-de-pacientes []
  (let [guilherme {:id 15, :nome "Guilherme"}
        daniela {:id 20, :nome "Daniela"}
        paulo {:id 25, :nome "Paulo"}

        ; uma variação com reduce, mais natural
        pacientes (reduce adiciona-paciente {} [guilherme, daniela, paulo])

        ; uma variação com shadowing, não é a melhor prática
        visitas {}
        visitas (adiciona-visitas visitas 15 ["01/01/2019"])
        visitas (adiciona-visitas visitas 20 ["01/02/2019", "01/01/2020"])
        visitas (adiciona-visitas visitas 15 ["01/03/2019"])]
    (pprint pacientes)
    (pprint visitas)

    ; não executa como deveria pois está sendo usado o simbolo paciente
    ; em vários lugares com significado diferente
    (imprime-relatorio-de-paciente visitas guilherme)
    (println (get visitas 20))))

(testa-uso-de-pacientes)

(pprint (s/validate Long 15))
;(pprint (s/validate Long "David"))
;(pprint (s/validate Long [15, 13]))

(s/set-fn-validation! true)

(s/defn teste-simples
  [x :- Long]
  (println x))

(teste-simples 30)
;(teste-simples "david")

(s/defn imprime-relatorio-de-paciente
  [visitas, paciente :- Long]
  (println "Visitas do paciente" paciente "são" (get visitas paciente)))

; Agora conseguimos o erro em tempo de *execução* que diz
; que o valor passado como parâmetro não condiz com o schema Long
;(testa-uso-de-pacientes)

(s/defn novo-paciente
  [id :- Long, nome :- s/Str]
  {:id id, :nome nome})

(pprint (novo-paciente 15 "David Borelli"))
;(pprint (novo-paciente "David Borelli" 15))

src/hospital_schemas/aula3.clj
(ns hospital_schemas.aula3
  (:use clojure.pprint)
  (:require [schema.core :as s]))

(s/set-fn-validation! true)

(def PosInt (s/pred pos-int? 'inteiro-positivo))

(def Paciente
  {:id PosInt :nome s/Str})

(s/defn novo-paciente :- Paciente
  [id :- PosInt
   nome :- s/Str]
  {:id id, :nome nome})

(pprint (novo-paciente 15 "David"))
;(pprint (novo-paciente -5 "David"))

(pprint "-----------------------------------------------------")

(defn maior-igual-zero?
  [valor]
  (>= valor 0))
(def ValorFinanceiro (s/constrained s/Num maior-igual-zero?))

(def Pedido
  {:paciente     Paciente
   :valor        ValorFinanceiro
   :procedimento s/Keyword})

; será que faz sentido "mini-schemas" como aliases ?
; (def Procimento s/Keyword)

(s/defn novo-pedido :- Pedido
  [paciente :- Paciente,
   valor :- ValorFinanceiro,
   procedimento :- s/Keyword]
  {:paciente paciente, :valor valor, :procedimento procedimento})

(pprint (novo-pedido (novo-paciente 15, "David Borelli"), 15.33, :raio-x))
;(pprint (novo-pedido (novo-paciente 15, "David Borelli"), -15.33, :raio-x))

(pprint "-----------------------------------------------------")

(def Numeros [s/Num])
(pprint (s/validate Numeros [15]))
(pprint (s/validate Numeros [15, 13]))
(pprint (s/validate Numeros [15, 13, 123, 132, 321, 12.23]))
(pprint (s/validate Numeros [0]))
; nil não é número, não faz sentido
; (pprint (s/validate Numeros [nil]))
(pprint (s/validate Numeros []))
(pprint (s/validate Numeros nil))

; nil não é s/Num
;(pprint (s/validate s/Num nil))
; nil é [s/Num]
(pprint (s/validate [s/Num] nil))

(def Plano [s/Keyword])
(pprint (s/validate Plano [:raio-x]))

(def Paciente
  {:id PosInt, :nome s/Str, :plano Plano})

(pprint (s/validate Paciente {:id 15, :nome "David", :plano [:raio-x, :ultrasom]}))
(pprint (s/validate Paciente {:id 15, :nome "David", :plano [:raio-x]}))
(pprint (s/validate Paciente {:id 15, :nome "David", :plano []}))
(pprint (s/validate Paciente {:id 15, :nome "David", :plano nil}))
; plano é uma keyword obrigatória no mapa, mas ela pode ter um valor vazio (nil)
; (pprint (s/validate Paciente {:id 15, :nome "David"}))















src/hospital_schemas/aula4.clj
(ns hospital_schemas.aula4
  (:use clojure.pprint)
  (:require [schema.core :as s]))

(s/set-fn-validation! true)

(def PosInt (s/pred pos-int? 'inteiro-positivo))
(def Plano [s/Keyword])
(def Paciente
  {:id                          PosInt,
   :nome                        s/Str,
   :plano                       Plano,
   (s/optional-key :nascimento) s/Str})

(pprint (s/validate Paciente {:id 15, :nome "David", :plano [:raio-x, :ultrasom]}))
(pprint (s/validate Paciente {:id 15, :nome "David", :plano [:raio-x]}))
(pprint (s/validate Paciente {:id 15, :nome "David", :plano []}))
(pprint (s/validate Paciente {:id 15, :nome "David", :plano [], :nascimento "27/09/1989"}))

; Outro tipo de uso de mapas
; Pacientes { 15 {PACIENTE}, 32 {PACIENTE} }
(def Pacientes
  {PosInt Paciente})

(pprint (s/validate Pacientes {}))
(let [david {:id 15, :nome "David", :plano [:raio-x]}
      erick {:id 20, :nome "Erick", :plano []}]
  (pprint (s/validate Pacientes {15 david}))
  (pprint (s/validate Pacientes {15 david, 20 erick}))
  ;(pprint (s/validate Pacientes {-15 david}))
  ;(pprint (s/validate Pacientes {15 15}))
  ;(pprint (s/validate Pacientes {15 {:id 15, :nome "David"}}))
  )

src/hospital_schemas/aula5.clj
(ns hospital_schemas.aula5
  (:use clojure.pprint)
  (:require [schema.core :as s]))

(s/set-fn-validation! true)

(def PosInt (s/pred pos-int? 'inteiro-positivo))
(def Plano [s/Keyword])
(def Paciente
  {:id                          PosInt,
   :nome                        s/Str,
   :plano                       Plano,
   (s/optional-key :nascimento) s/Str})

(def Pacientes
  {PosInt Paciente})

(def Visitas
  {PosInt [s/Str]})


; Garantia que existe um ID válido
; (pprint (s/validate Paciente { :id nil, :nome "David", :plano [] }))

; Foi removido o if-def, também o throw pois
; o schema GARANTIU a existência do ID e a VALIDADE do id
; ... se a validação estiver ativa
(s/defn adiciona-paciente :- Pacientes
  [pacientes :- Pacientes, paciente :- Paciente]
  (let [id (:id paciente)]
    (assoc pacientes id paciente)))

(s/defn adiciona-visitas :- Visitas
  [visitas :- Visitas, paciente :- PosInt, novas-visitas :- [s/Str]]
  (if (contains? visitas paciente)
    (update visitas paciente concat novas-visitas)
    (assoc visitas paciente novas-visitas)))

(s/defn imprime-relatorio-de-paciente [visitas :- Visitas, paciente :- PosInt]
  (println "Visitas do paciente" paciente "são" (get visitas paciente)))

(defn testa-uso-de-pacientes []
  (let [guilherme {:id 15, :nome "Guilherme", :plano []}
        daniela {:id 20, :nome "Daniela", :plano []}
        paulo {:id 25, :nome "Paulo", :plano []}

        ; uma variação com reduce, mais natural
        pacientes (reduce adiciona-paciente {} [guilherme, daniela, paulo])

        ; uma variação com shadowing, não é a melhor prática
        visitas {}
        visitas (adiciona-visitas visitas 15 ["01/01/2019"])
        visitas (adiciona-visitas visitas 20 ["01/02/2019", "01/01/2020"])
        visitas (adiciona-visitas visitas 15 ["01/03/2019"])]
    (pprint pacientes)
    (pprint visitas)

    (imprime-relatorio-de-paciente visitas 20)))

(testa-uso-de-pacientes)

src/hospital_schemas/aula2.clj
(ns hospital_schemas.aula2
  (:use clojure.pprint)
  (:require [schema.core :as s]))

(s/set-fn-validation! true)

;(s/defrecord Paciente
;  [id :- Long, nome :- s/Str])

;(map->Paciente {15 "David"}) Baseado em um mapa
;(->Paciente 15 "David") Para criar um Paciente

;(pprint (map->Paciente {25 "David Borelli"}))
;(pprint (map->Paciente {"25" "David Borelli"}))

(def Paciente
  "Schema de um paciente"
  {:id s/Num, :nome s/Str})

(pprint (s/explain Paciente))
(pprint (s/validate Paciente {:id 15, :nome "guilherme"}))

; Typo é pego pelo schema, mas poderiamos argumentas que esse
; tipo de erro seria pego em testes automatizados com uma boa cobertura

;(pprint (s/validate Paciente { :id 15, :name "guilherme" }))

; MAS... entra a questão de QUERER ser forward compatible OU NÃO.
; entender esse trade-off
; sistemas externos não me quebraram ao adicionar novos campos (forward compatible)
; no nosso validate não estamos sendo forward compatible (pode ser interessante quando queremos analisar mudanças)
;(pprint (s/validate Paciente { :id 15, :nome "guilherme", :plano [:raio-x] }))



; Chaves que são keywords em schemas são por padrão OBRIGATÓRIAS
;(pprint (s/validate Paciente { :id 15 }))


; Tipo de retorno com schema
; força a validação na saída da função
;(s/defn novo-paciente :- Paciente
;  [id :- s/Num, nome :- s/Str]
;  {:id id, :nome nome, :plano []})

(s/defn novo-paciente :- Paciente
  [id :- s/Num, nome :- s/Str]
  {:id id, :nome nome})

(pprint (novo-paciente 15 "David Borelli"))



; Função pura simples, fácil de testar
(defn estritamente-positivo? [x]
  (> x 0))
(def EstritamentePositivo (s/pred estritamente-positivo? 'estritamente-positivo))
;(def EstritamentePositivo (s/pred estritamente-positivo?))

(pprint (s/validate EstritamentePositivo 15))
;(pprint (s/validate EstritamentePositivo 0))
;(pprint (s/validate EstritamentePositivo -15))

(def Paciente
  "Schema de um paciente"
  { :id (s/constrained s/Int pos?), :nome s/Str })
; é por isso que é importante conhecer bem a documentação
; já existe pos? e tbm pos-int implementado
; UMA DICA: Sempre olhar documentação

(pprint (s/validate Paciente { :id 15, :nome "Guilherme" }))
;(pprint (s/validate Paciente { :id -15, :nome "Guilherme" }))
;(pprint (s/validate Paciente { :id 0, :nome "Guilherme" }))

(def Paciente
  "Schema de um paciente"
  { :id (s/constrained s/Int #(> % 0) 'inteiro-estritamente-possitivo), :nome s/Str })
; Usar lambida não é tão interessante pois na mensagem de erro, é apresentado um nome
; aleatório de função anônima, é bom evitar, parece uma péssima prática

(pprint (s/validate Paciente { :id 1, :nome "David" }))